### DISCUSSION TITLE
Improvig the sympy.printing.pretty implementation

### DISCUSSION THREAD

**Discussion by mmatera** [sympy/sympy#27264] [Ideas]
In #27257, I proposed a set of minimal changes to this module to make it usable to build PrettyPrint-like equations from objects that are not Sympy objects (mathics-core expressions, for example). To ensure getting the same look, it would be helpful to provide `stringPict` objects with a set of methods to draw primitives, and then use them along all the methods in `PrettyPrinter`.

Looking at the code, and from the discussion started in #27257, it becomes clear that apart from the minimal changes I had proposed, the module would require a refactor, driving it closer to current Python's code quality standards and conventions, removing duplicated code, and to define clearer abstraction layers.

I open this discussion to see if we can go deeper into the proposal for that refactor, assuming there is interest in the community of Sympy maintainers. 

If that is the case, I prepared a new private branch https://github.com/mmatera/sympy/tree/prettyform-black  
which, avoiding breaking compatibility, applies minimal changes to fulfill the current standards, and to sort the methods and objects in the package in lexicographical order.

**Comment by oscarbenjamin** [comment]
In general we have decided not to use black after exploring what it does to some code (gh-22434). In this case it makes the diff much bigger than it should be which makes it hard to see exactly what the meaningful changes are.

The pretty printing does need a refactor but not because of code formatting. The problems are in the general design of how all the `_print_X` methods are used for each type when it would be better to transform each printed expression into a display form expression and then have a smaller number of methods for actually rendering the display of different types of objects. What that means is that most of the code should probably end up being replaced by new code.

**Reply by mmatera** [reply]
> In general we have decided not to use black after exploring what it does to some code ([gh-22434](https://github.com/sympy/sympy/pull/22434)). In this case it makes the diff much bigger than it should be which makes it hard to see exactly what the meaningful changes are.
> 

OK, but I found it useful to sort methods in lexicographical order. On the other hand, https://github.com/mmatera/sympy/tree/prettyform-black does not have any *meaningful* change: It is just about putting the existent code in order, fixing conventions (the ones that spyder's code analyzer reports as warnings) and hiding functions that are not part of the interface, but just helpers to initialize a module.  

> The pretty printing does need a refactor but not because of code formatting. 

To start thinking about such a refactor, I thought it would be useful to start with a tidier version of the current code, with a clearer division of what piece has each functionality, what is supposed to be part of the public interface, and what are the internal functions. For example, in `sympy.printing.pretty.pretty_symbology` there are a lot of functions that does not seems to have any other functionality than initializing some dictionaries, like `greek_unicode` or `fraction`. These functions, that nobody worried about looking for a descriptive name, or adding a docstring, should be internal functions.

> The problems are in the general design of how all the `_print_X` methods are used for each type when it would be better to transform each printed expression into a display form expression and then have a smaller number of methods for actually rendering the display of different types of objects. What that means is that most of the code should probably end up being replaced by new code.

My guess is that once certain "rendering" primitives are identified ( put two stringPicts together in certain relative positions, drawing certain multicharacter symbols, adding parenthesis, building grids/tables of elements) much of the `_print_*` methods can be reduced and generalized. Then, the "intermediate" algebraic->formatting stage should become clearer.

**Reply by oscarbenjamin** [reply]
> These functions, that nobody worried about looking for a descriptive name, or adding a docstring, should be internal functions.

They are internal functions. I don't think that the changes to pretty_symbology are really an improvement though.

I realise that editors like spyder may complain about lots of things but that is because it is using some linter and we do have linting in sympy using `ruff` but lots of particular lint rules are turned off because we don't want to mandate them in the codebase. A lint rule that requires code to be ordered lexicographically would most certainly not be used.

New code can be written in a different style. When there is a reason to change code then it can be written in a different style as well. It is better not to bulk reformat long-standing code though unless we are going to agree to do that on a project-wide basis.

**Reply by mmatera** [reply]
> > These functions, that nobody worried about looking for a descriptive name, or adding a docstring, should be internal functions.
> They are internal functions. I don't think that the changes to pretty_symbology are really an improvement though.
> 

OK, but in order to make a refactor, it would be nice to know what are internal and what not. Notice for example that functions like `G` appears in the sphinx documentation.

> I realise that editors like spyder may complain about lots of things but that is because it is using some linter and we do have linting in sympy using `ruff` but lots of particular lint rules are turned off because we don't want to mandate them in the codebase. A lint rule that requires code to be ordered lexicographically would most certainly not be used.
> 
> New code can be written in a different style. When there is a reason to change code then it can be written in a different style as well. It is better not to bulk reformat long-standing code though unless we are going to agree to do that on a project-wide basis.

OK, but are there criteria for ordering methods and objects in classes and modules across the whole project?

**Reply by oscarbenjamin** [reply]
> are there criteria for ordering methods and objects in classes and modules across the whole project?

No. Usually e.g. methods of a class are grouped or ordered by having related functionality.

> Notice for example that functions like `G` appears in the sphinx documentation.

I'm not sure why they are there. I would generally consider the whole `pretty_symbology` module to be internal. Is there a need to make significant changes to it though?

**Reply by mmatera** [reply]
> > are there criteria for ordering methods and objects in classes and modules across the whole project?
> 
> No. Usually e.g. methods of a class are grouped or ordered by having related functionality.

OK

> > Notice for example that functions like `G` appears in the sphinx documentation.
> 
> I'm not sure why they are there. I would generally consider the whole `pretty_symbology` module to be internal. Is there a need to make significant changes to it though?

Not too much. Just remove them from the sphix documentation,  (see changes here: https://github.com/mmatera/sympy/tree/prettyform-black) The other changes in `pretty_symbology`  are just to put all the initialization code in an initialization function, and sorting the code. In this case, the lexicographic order and functionality order are more or less the same.
